hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/HoodieTimelineArchiver.java

```diff
@@ -388,60 +387,40 @@ private Map<String, Boolean> deleteFilesParallelize(HoodieTableMetaClient metaCl
-        .map(hoodieInstants -> {
-          if (hoodieInstants.size() > this.maxInstantsToKeep) {
-            return hoodieInstants.subList(0, hoodieInstants.size() - this.minInstantsToKeep);
-          } else {
-            return new ArrayList<HoodieInstant>();
-          }
-        }).flatMap(Collection::stream);
-  }
-
-  private Stream<HoodieInstant> getCommitInstantsToArchive() throws IOException {
-    // TODO (na) : Add a way to return actions associated with a timeline and then merge/unify
-    // with logic above to avoid Stream.concat
-    HoodieTimeline commitTimeline = table.getCompletedCommitsTimeline();
-
-    // Get the oldest inflight instant and a completed commit before this inflight instant.
-    Option<HoodieInstant> oldestPendingInstant = table.getActiveTimeline()
-        .getWriteTimeline()
-        .filter(instant -> !instant.isCompleted())
-        .firstInstant();
-
-    // Oldest commit to retain is the greatest completed commit, that is less than the oldest pending instant.
-    // In some cases when inflight is the lowest commit then oldest commit to retain will be equal to oldest
-    // inflight commit.
-    Option<HoodieInstant> oldestCommitToRetain;
-    if (oldestPendingInstant.isPresent()) {
-      Option<HoodieInstant> completedCommitBeforeOldestPendingInstant =
-          Option.fromJavaOptional(commitTimeline.getReverseOrderedInstants()
-              .filter(instant -> HoodieTimeline.compareTimestamps(instant.getTimestamp(),
-                  LESSER_THAN, oldestPendingInstant.get().getTimestamp())).findFirst());
-      // Check if the completed instant is higher than the oldest inflight instant
-      // in that case update the oldestCommitToRetain to oldestInflight commit time.
-      if (!completedCommitBeforeOldestPendingInstant.isPresent()
-          || HoodieTimeline.compareTimestamps(oldestPendingInstant.get().getTimestamp(),
-          LESSER_THAN, completedCommitBeforeOldestPendingInstant.get().getTimestamp())) {
-        oldestCommitToRetain = oldestPendingInstant;
+  private List<HoodieInstant> getCommitInstantsToArchive() throws IOException {
+    HoodieTimeline completedCommitsTimeline = table.getCompletedCommitsTimeline();
+
+    if (completedCommitsTimeline.countInstants() > maxInstantsToKeep) {
+      // Step1: Get all candidates of oldestInstantToRetain.
+      List<Option<HoodieInstant>> oldestInstantToRetainCandidates = new ArrayList<>();
+
+      // 1. Oldest commit to retain is the greatest completed commit, that is less than the oldest pending instant.
+      // In some cases when inflight is the lowest commit then oldest commit to retain will be equal to the oldest
+      // inflight commit.
+      // By filtering out any commit >= oldestCommitToRetain, we can ensure there are no gaps in the timeline
+      // when inflight commits are present.
+      Option<HoodieInstant> oldestPendingInstant = table.getActiveTimeline()
```

Comment:
nit: for naming convention of the variables and docs, shall we use the `earliest` (already used in existing code) instead of the `oldest` and the `latest` instead of the `newest` for consistency?

Reply:
Not strong preference, `earliest` and `latest` sounds better.

Reply:
Yeah, just ensure the terms are consistent in the code and docs (we should not have both `earliest` and `oldest`).

Reply:
> Yeah, just ensure the terms are consistent in the code and docs (we should not have both `earliest` and `oldest`).

I checked, there are 29 oldest and 7 Earliest in `HoodieTimelineArchiver`. And there are plenty of oldest elsewhere. The changes is huge, it would not be good to modify it in this PR, what do you think?

Reply:
For this PR, let's make sure the terms are consistent in `HoodieTimelineArchiver`, not considering other classes.

Reply:
So, do we need to change this? In fact, in `HoodieTimelineArchiver`, there are many more oldest than earliest
```java
ClusteringUtils.getOldestInstantToRetainForClustering(...)
CompactionUtils.getOldestInstantToRetainForCompaction(...)
```

Reply:
Just rename these methods should be fine.